lgc/builder/Builder.cpp

```diff
@@ -256,55 +262,40 @@ PointerType *Builder::getBufferDescTy(Type *pointeeTy) {
-}
-
-// =====================================================================================================================
-// Get the type of an fmask descriptor
-VectorType *Builder::getFmaskDescTy() {
-  return FixedVectorType::get(getInt32Ty(), 8);
-}
-
-// =====================================================================================================================
-// Get the type of a texel buffer descriptor
-VectorType *Builder::getTexelBufferDescTy() {
-  return FixedVectorType::get(getInt32Ty(), 4);
-}
-
-// =====================================================================================================================
-// Get the type of a sampler descriptor
-VectorType *Builder::getSamplerDescTy() {
-  return FixedVectorType::get(getInt32Ty(), 4);
-}
-
-// =====================================================================================================================
-// Get the type of pointer to image descriptor.
-// This is in fact a struct containing the pointer itself plus the stride in dwords.
-Type *Builder::getImageDescPtrTy() {
-  return StructType::get(getContext(), {PointerType::get(getImageDescTy(), ADDR_SPACE_CONST), getInt32Ty()});
-}
-
-// =====================================================================================================================
-// Get the type of pointer to fmask descriptor.
-// This is in fact a struct containing the pointer itself plus the stride in dwords.
-Type *Builder::getFmaskDescPtrTy() {
-  return StructType::get(getContext(), {PointerType::get(getFmaskDescTy(), ADDR_SPACE_CONST), getInt32Ty()});
-}
-
-// =====================================================================================================================
-// Get the type of pointer to texel buffer descriptor.
-// This is in fact a struct containing the pointer itself plus the stride in dwords.
-Type *Builder::getTexelBufferDescPtrTy() {
-  return StructType::get(getContext(), {PointerType::get(getTexelBufferDescTy(), ADDR_SPACE_CONST), getInt32Ty()});
+// Get the type of a descriptor
+//
+// @param descType : Descriptor type, one of the ResourceNodeType values
+VectorType *Builder::getDescTy(ResourceNodeType descType) {
+  const GpuProperty &gpuProperty = getLgcContext()->getTargetInfo().getGpuProperty();
+  unsigned byteSize = 0;
+  switch (descType) {
+  case ResourceNodeType::DescriptorBuffer:
+  case ResourceNodeType::DescriptorTexelBuffer:
+    byteSize = gpuProperty.descriptorSizeBuffer;
+    break;
+  case ResourceNodeType::DescriptorBufferCompact:
+    byteSize = DescriptorSizeBufferCompact;
+    break;
+  case ResourceNodeType::DescriptorSampler:
+    byteSize = gpuProperty.descriptorSizeSampler;
+    break;
+  case ResourceNodeType::DescriptorResource:
+  case ResourceNodeType::DescriptorFmask:
+    byteSize = gpuProperty.descriptorSizeResource;
+    break;
+  default:
+    llvm_unreachable("");
+    break;
+  }
+  return FixedVectorType::get(getInt32Ty(), byteSize / sizeof(uint32_t));
```

Comment:
Have you considered creating these types once during constructions and just returning that?

Reply:
Well, maybe, but I don't really like the idea of second-guessing what caching (if any) FixedVectorType has.

Reply:
That's fair.